x/concentrated-liquidity/lp_test.go

```diff
@@ -622,6 +622,326 @@ func (s *KeeperTestSuite) TestWithdrawPosition() {
+			setupConfig: baseCase,
+
+			// system under test parameters
+			sutConfigOverwrite: &lpTest{
+				amount0Expected: amount0PerfectRatio.Add(amount0PerfectRatio),
+				amount1Expected: amount1PerfectRatio.Add(amount1PerfectRatio),
+			},
+			timeElapsed: defaultTimeElapsed,
+			// We add back in the rounding error for this test case to demonstrate that rounding pushes us off the boundary
+			// in the previous test case
+			amount0ToAdd: amount0PerfectRatio.Add(roundingError),
+			amount1ToAdd: amount1PerfectRatio,
+		},
+		"add partial liquidity amount": {
+			// setup parameters for creating a pool and position.
+			setupConfig: baseCase,
+
+			// system under test parameters
+			sutConfigOverwrite: &lpTest{
+				amount0Expected: amount0PerfectRatio.Add(amount0PerfectRatio.QuoRaw(2)).Sub(roundingError),
+				// Since we round on the other the asset when we withdraw, asset0 turns into the bottleneck and
+				// thus we cannot use the full amount of asset1. We calculate the output here (rounding up):
+				// https://www.wolframalpha.com/input?i=2276822106.317042173055221520+*+%2870.710678118654752440+-+67.416615162732695594%29
+				amount1Expected: sdk.NewInt(7499995358),
+			},
+			timeElapsed:  defaultTimeElapsed,
+			amount0ToAdd: amount0PerfectRatio.QuoRaw(2),
+			amount1ToAdd: amount1PerfectRatio.QuoRaw(2),
+		},
+		"add partial liquidity amount (with rounding error added back in)": {
+			// setup parameters for creating a pool and position.
+			setupConfig: baseCase,
+
+			// system under test parameters
+			sutConfigOverwrite: &lpTest{
+				amount0Expected: amount0PerfectRatio.Add(amount0PerfectRatio.QuoRaw(2)),
+				amount1Expected: amount1PerfectRatio.Add(amount1PerfectRatio.QuoRaw(2)),
+			},
+			timeElapsed:  defaultTimeElapsed,
+			amount0ToAdd: amount0PerfectRatio.QuoRaw(2).Add(roundingError),
+			amount1ToAdd: amount1PerfectRatio.QuoRaw(2),
+		},
+
+		// Cases for adding to superfluid staked positions
+		// TODO: add support for adding to superfluid positions so these are not error cases
+
+		"error: attempt to add to a position with underlying lock that has finished unlocking": {
+			// setup parameters for creating a pool and position.
+			setupConfig: baseCase,
+
+			// system under test parameters
+			sutConfigOverwrite: &lpTest{
+				amount0Expected: amount0PerfectRatio.Add(amount0PerfectRatio).Sub(roundingError),
+				// Since we round on the other the asset when we withdraw, asset0 turns into the bottleneck and
+				// thus we cannot use the full amount of asset1. We calculate the output here (rounding up):
+				// https://www.wolframalpha.com/input?i=3035764327.860030912175533748+*+%2870.710678118654752440+-+67.416615162732695594%29
+				amount1Expected: sdk.NewInt(9999998816),
+				expectedError:   types.PositionSuperfluidStakedError{PositionId: uint64(1)},
+			},
+			timeElapsed:  defaultTimeElapsed,
+			amount0ToAdd: amount0PerfectRatio,
+			amount1ToAdd: amount1PerfectRatio,
+
+			createLockUnlocked: true,
```

Comment:
If the underlying lock has finished unlocking, then shouldn't we update the underlyinglock field accordingly and therefore allow to add to position with no error?

Reply:
I figured all lock-related logic would fall under the superfluid `AddToPosition` PR – I did not want to step on your toes by implementing some of it piecewise. Happy to add this in if you'd like though

Reply:
Sorry for the late response. Just to be clear, zero logic from addToPosition for superfluid overlaps with this PR. So you can treat any attempt to addToPosition with an underlying lock as an error, while any position that has completed unlocking should be going through this flow if that makes sense.